lib/BuildSystem/BuildSystem.cpp

```diff
@@ -2430,6 +2437,28 @@ class StaleFileRemovalCommand : public Command {
     bsci.getDelegate().commandStarted(this);
 
     for (auto fileToDelete : filesToDelete) {
+      // If no root paths are specified, any path is valid.
+      bool isLocatedUnderRootPath = roots.size() == 0 ? true : false;
+
+      // If root paths are defined, stale file paths should be absolute.
+      if (roots.size() > 0 && fileToDelete[0] != '/') {
+        bsci.getDelegate().error("", {}, (Twine("Stale file '") + fileToDelete + "' has a relative path. This is invalid in combination with the root path attribute."));
+        continue;
+      }
+
+      // Check if the file is located under one of the allowed root paths.
+      for (auto root : roots) {
+        auto res = std::mismatch(root.begin(), root.end(), fileToDelete.begin());
+        if (res.first == root.end()) {
+          isLocatedUnderRootPath = true;
+        }
```

Comment:
This should also check that the next character in the `fileToDelete` is a path separator, right?